Added bigfile tests

BobDickinson opened the pull request:
Implemented previously stubbed out large/raw file download tests for all providers except Amazon and Google.

BobDickinson commented:
I implemented the mock support for raw file download and unstubbed the tests for all except Amazon and Google, because neither of them support the download callback param.  The documentation for file download indicates than an error callback is supported:

    client.download(options, function (err) { })

The tests that were stubbed out had a different expectation, namely a callback called on error *or* completion, that when called on completion, supplies a file object (containing at least a name and size).  All providers that support the callback call it in this way, and the test code relies on that.

At very least, the Amazon and Google providers need to support the callback on error.  If the intent is that the download callback provides the file object on completion, then the Amazon and Google providers also need to be updated to support that (and the docs updated).  To do this on Google is going to require an update to gcloud, which has it's own set of challenges (will open an issue on that).

This PR could be merged as-is, in support of #469 (all of the providers affected by that PR are tested here - and work both with and without that fix).  In that case, we should probably open issues on the lack of callback support for Amazon and Google.

But given that I ran into two failing providers, I didn't just want to check this in without some review.

kenperkins commented:
It definitely appears that the large file download should be updated to behave based on the events strategy instead of the callback strategy.

BobDickinson commented:
I'll change the big file tests to just read the stream and not rely on the completion callback.

As far as I can tell, the `_request`-based download calls still need to support the error callback (for cases where the error condition occurs before you have produced the read stream).  But after looking at the source for the AWS and gcloud libs, I think you're guaranteed to get a read stream back and get your errors as events on that read stream for those (so I guess they're fine as-is).

BobDickinson commented:
OK, finally.  Sorry this was such a trainwreck.  Still working on my process (4 of the  commits were to fix issues that worked in my env and failed in travis).  I think I'm good if I just remember to test on 0.10.x and 4.x and always use `make travis`.

I converted the big/raw download tests to only rely on the stream and stream events.  

The Amazon provider has a different behavior and I'd like someone to weigh-in on my workaround.  When you call `stream.end()` on the stream you get back from Amazon, it does not flush the stream and then trigger the "end" event (as it does with the other providers).  It triggers the "end" event immediately, then flushes the stream.  So you get an error out of AWS trying to write after close.  I found that if I actually consume the stream from Amazon (pipe it to null), then it works as expected, which I think means it works fine in the wild also.  But I'm not really a master of streams.  Would like a second opinion.

kenperkins commented:
First off, none of this is a trainwreck; It's a trainwreck because I've done a poor job of getting the streaming interface perfectly implemented for both upload and download, and you're just dealing with the fallout.

I swore I had implemented the proper event handling for all providers, but clearly that's not working. That said, I don't think that impacts your tests, just the broader features for streaming.

kenperkins commented:
So given our discovery in #469 about the callback nature of the issue, these tests won't ensure that we won't still break, but are good to have none-the-less. :+1: to merge.

BobDickinson commented:
You want failing tests?  That's my specialty!

Added tests with 300Mb file that fail.  If you apply the fix in #469, plus tweak Azure download to use null encoding, these tests should pass.

kenperkins commented on test/common/storage/base-test.js:
```diff
@@ -22,12 +22,26 @@ var fs = require('fs'),
+for (var i = 0; i < buff.length; i++) {
+  buff[i] = i % 256;
+}
+fs.write(fd, buff, 0, buff.length, 0);
```
I'm wondering if we even need to create the file on disk, versus just using a stream that we randomly send bits out of until we have sent enough.

BobDickinson commented:
I rewrote the tests to stream random data, but it turns out Hock doesn't support the stream response mode.  Nock does support that (if the object you pass to `reply` is a stream, it does the right thing --  `replyWithFile` just opens a read stream on the file and calls `reply` for example).   But Hock did not copy that particular feature.  If you pass a non-string to `reply` in Hock, it just does a JSON.stringify on it.  You can't use a string that large, as it exceeds the string size.  And if you try to pass Hock a Buffer of that size, it explodes (it doesn't know anything about buffers anyway).

So I added stream support to Hock `reply` - see: https://github.com/mmalecki/hock/pull/24.  A side effect is that it now also handles Buffers appropriately. 

If Hock (@kenperkins) takes my PR and publishes to npm, then I'd be happy to update pkgcloud to use that version of Hock and check in my dynamic random streaming version of the 300Mb tests instead of the generated-disk-file-version.

kenperkins commented:
I'll get on this after my morning meetings and :coffee:

kenperkins commented:
@BobDickinson published `hock@1.3.0`.

BobDickinson commented:
Per recent comments on #469, these large stream download tests are not going to work using the download\request callback on Node 0.10.x.  The checked-in version tests download of 300Mb of randomly generated data without using the callback. Given that, these tests don't really test anything new, and have no relation to #469, but I guess they don't hurt anything.
